samples/sample1/billing_queue.h

```diff
@@ -0,0 +1,93 @@
+/*
+    Copyright 2017-2018 Dynatrace LLC
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#ifndef SAMPLE1_BILLING_QUEUE_H_INCLUDED
+#define SAMPLE1_BILLING_QUEUE_H_INCLUDED
+
+#include <exception>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+#include <mutex>
+#include <condition_variable>
+#include <queue>
+
+/*========================================================================================================================================*/
+
+struct billable_usage {
+    unsigned changed_chars;
+    unsigned total_chars;
+};
+
+class message_queue {
+public:
+
+    // For reliable messaging, as you would want for billing, some sort of
+    // "processed" flag would be required that has to be set for the message
+    // so that messages which were sucessfully received but where processing
+    // failed can be redelivered.
+    struct queue_message {
+        std::unordered_map<std::string, std::string> headers;
+        std::string message_id;
+        billable_usage payload;
+
+        void mark_processed() noexcept {} // Not implemented.
+    };
+
+    message_queue(message_queue const&) = delete; // We're non-copyable.
+    message_queue operator =(message_queue const&) = delete; // We're non-copyable.
+
+    message_queue() = default;
+
+    void send(queue_message& message) {
+        message.message_id = std::to_string(m_queue.size());
+        try {
+            std::lock_guard<std::mutex> lock(m_queue_mut);
+            m_queue.push(message);
+        } catch (...) {
+            message.message_id.clear();
+            throw;
+        }
+    }
+
+    queue_message poll_receive_one() {
+        std::unique_lock<std::mutex> lock(m_queue_mut);
+        if (m_queue.empty())
+            return {};
+        auto result = std::move(m_queue.front());
+        m_queue.pop();
+        return result;
+    }
+
+private:
+    std::mutex m_queue_mut;
+    std::queue<queue_message> m_queue;
+};
+
+char const* const BillingQueueName = "billing-queue";
+
+inline message_queue& connect_queue(std::string const& queue_name) { // Not thread safe when creating any queues!
```

Comment:
Since the sample requires C++11 we could just as well make this thread-safe. Locking a `std::mutex` here isn't really more complicated than writing that "not thread safe" comment.